ddl/job_table.go

```diff
@@ -0,0 +1,564 @@
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	addingDDLJobGeneral = "/tidb/ddl/add_ddl_job_general"
+	addingDDLJobReorg   = "/tidb/ddl/add_ddl_job_reorg"
+)
+
+func (dc *ddlCtx) insertRunningDDLJobMap(id int64) {
+	dc.runningJobs.Lock()
+	defer dc.runningJobs.Unlock()
+	dc.runningJobs.ids[id] = struct{}{}
+}
+
+func (dc *ddlCtx) deleteRunningDDLJobMap(id int64) {
+	dc.runningJobs.Lock()
+	defer dc.runningJobs.Unlock()
+	delete(dc.runningJobs.ids, id)
+}
+
+func (dc *ddlCtx) excludeJobIDs() string {
+	dc.runningJobs.RLock()
+	defer dc.runningJobs.RUnlock()
+	if len(dc.runningJobs.ids) == 0 {
+		return ""
+	}
+	dc.runningJobIDs = dc.runningJobIDs[:0]
+	for id := range dc.runningJobs.ids {
+		dc.runningJobIDs = append(dc.runningJobIDs, strconv.Itoa(int(id)))
+	}
+	return fmt.Sprintf("and job_id not in (%s)", strings.Join(dc.runningJobIDs, ","))
+}
+
+const (
+	getJobSQL = "select job_meta, processing from mysql.tidb_ddl_job where job_id in (select min(job_id) from mysql.tidb_ddl_job group by schema_ids, table_ids) and %s reorg %s order by processing desc, job_id"
+)
+
+type jobType int
+
+func (t jobType) String() string {
+	switch t {
+	case general:
+		return "general"
+	case reorg:
+		return "reorg"
+	}
+	return "unknown job type: " + strconv.Itoa(int(t))
+}
+
+const (
+	general jobType = iota
+	reorg
+)
+
+func (d *ddl) getJob(sess *session, tp jobType, filter func(*model.Job) (bool, error)) (*model.Job, error) {
+	not := "not"
+	label := "get_job_general"
+	if tp == reorg {
+		not = ""
+		label = "get_job_reorg"
+	}
+	sql := fmt.Sprintf(getJobSQL, not, d.excludeJobIDs())
+	rows, err := sess.execute(context.Background(), sql, label)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, row := range rows {
+		jobBinary := row.GetBytes(0)
+		runJob := model.Job{}
+		err := runJob.Decode(jobBinary)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if row.GetInt64(1) == 1 {
+			return &runJob, nil
+		}
+		b, err := filter(&runJob)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if b {
+			if err := d.markJobProcessing(sess, &runJob); err != nil {
+				logutil.BgLogger().Warn("[ddl] handle ddl job failed: mark job is processing meet error", zap.Error(err), zap.String("job", runJob.String()))
+				return nil, errors.Trace(err)
+			}
+			return &runJob, nil
+		}
+	}
+	return nil, nil
+}
+
+func (d *ddl) getGeneralJob(sess *session) (*model.Job, error) {
+	return d.getJob(sess, general, func(job *model.Job) (bool, error) {
+		if job.Type == model.ActionDropSchema {
+			sql := fmt.Sprintf("select job_id from mysql.tidb_ddl_job where find_in_set(%s, schema_ids) != 0 and processing limit 1", strconv.Quote(strconv.FormatInt(job.SchemaID, 10)))
+			return d.checkJobIsRunnable(sess, sql)
+		}
+		sql := fmt.Sprintf("select job_id from mysql.tidb_ddl_job t1, (select table_ids from mysql.tidb_ddl_job where job_id = %d) t2 where processing and find_in_set(t1.table_ids, t2.table_ids) != 0", job.ID)
```

Comment:
Are the following results problematic for multi-schema operations？
```
mysql> select FIND_IN_SET('3', '1,2,3');
+---------------------------+
| FIND_IN_SET('3', '1,2,3') |
+---------------------------+
|                         3 |
+---------------------------+
1 row in set (0.00 sec)

mysql> select FIND_IN_SET('1,3', '1,2,3');
+-----------------------------+
| FIND_IN_SET('1,3', '1,2,3') |
+-----------------------------+
|                           0 |
+-----------------------------+
1 row in set (0.00 sec)
```

Reply:
the general job can only be concurrency with reorg job, but reorg job does not contain multiple tables or schemas operation